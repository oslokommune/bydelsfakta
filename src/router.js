import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
