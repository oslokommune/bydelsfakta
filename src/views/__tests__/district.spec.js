import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import VueMeta from 'vue-meta';

import Bydel from '../District';
import setupI18n from '../../i18n';
import { routes } from '../../router';
import mockStore from '../../../tests/MockStore';

const i18n = setupI18n();

describe('Bydel', () => {
  let wrapper = null;
  let router = null;
  let store = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    localVue.use(Vuex);
    localVue.use(VueMeta);
    router = new VueRouter({ routes });
    store = new Vuex.Store(mockStore);
    wrapper = mount(Bydel, {
      propsData: {
        district: 'gamleoslo',
      },
      localVue,
      router,
      store,
      i18n,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders bydel-component and finds main-container class', () => {
    expect(wrapper.classes('main-container')).toBe(true);
  });

  test('renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
