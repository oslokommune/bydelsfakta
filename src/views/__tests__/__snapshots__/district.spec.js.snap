// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bydel renders component correctly 1`] = `
<div
  class="main-container"
>
  <div
    class="main-container__cards main-container__cards--district"
  >
    <div
      class="topics-grid"
    >
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/befolkningsutvikling"
        id="befolkningsutvikling"
      >
        <picture>
          <source
            alt="Befolkningsutvikling"
            srcset="/img/folkemengde.webp"
            type="image/webp"
          />
           
          <source
            alt="Befolkningsutvikling"
            srcset="/img/folkemengde.png"
            type="image/png"
          />
           
          <img
            alt="Befolkningsutvikling"
            aria-hidden="true"
            class="oslo__img"
            src="/img/folkemengde.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(245, 173, 165);"
        >
          Befolkning
        </span>
         
        <span
          class="oslo__topic"
        >
          Befolkningsutvikling
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/fodte-dode-flytting"
        id="fodte-dode-flytting"
      >
        <picture>
          <source
            alt="Fødte, døde og flytting"
            srcset="/img/flytting.webp"
            type="image/webp"
          />
           
          <source
            alt="Fødte, døde og flytting"
            srcset="/img/flytting.png"
            type="image/png"
          />
           
          <img
            alt="Fødte, døde og flytting"
            aria-hidden="true"
            class="oslo__img"
            src="/img/flytting.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(245, 173, 165);"
        >
          Befolkning
        </span>
         
        <span
          class="oslo__topic"
        >
          Fødte, døde og flytting
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/alder"
        id="alder"
      >
        <picture>
          <source
            alt="Kjønn og alder"
            srcset="/img/alder.webp"
            type="image/webp"
          />
           
          <source
            alt="Kjønn og alder"
            srcset="/img/alder.png"
            type="image/png"
          />
           
          <img
            alt="Kjønn og alder"
            aria-hidden="true"
            class="oslo__img"
            src="/img/alder.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(245, 173, 165);"
        >
          Befolkning
        </span>
         
        <span
          class="oslo__topic"
        >
          Kjønn og alder
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/innvandrerbefolkningen"
        id="innvandrerbefolkningen"
      >
        <picture>
          <source
            alt="Personer med innvandrer&shy;bakgrunn"
            srcset="/img/innvandring.webp"
            type="image/webp"
          />
           
          <source
            alt="Personer med innvandrer&shy;bakgrunn"
            srcset="/img/innvandring.png"
            type="image/png"
          />
           
          <img
            alt="Personer med innvandrer&shy;bakgrunn"
            aria-hidden="true"
            class="oslo__img"
            src="/img/innvandring.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(245, 173, 165);"
        >
          Befolkning
        </span>
         
        <span
          class="oslo__topic"
        >
          Personer med innvandrer­bakgrunn
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/husholdninger"
        id="husholdninger"
      >
        <picture>
          <source
            alt="Husholdninger"
            srcset="/img/husholdninger.webp"
            type="image/webp"
          />
           
          <source
            alt="Husholdninger"
            srcset="/img/husholdninger.png"
            type="image/png"
          />
           
          <img
            alt="Husholdninger"
            aria-hidden="true"
            class="oslo__img"
            src="/img/husholdninger.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(245, 173, 165);"
        >
          Befolkning
        </span>
         
        <span
          class="oslo__topic"
        >
          Husholdninger
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/levekaar"
        id="levekaar"
      >
        <picture>
          <source
            alt="Levekårs&shy;indikatorer"
            srcset="/img/levekaar.webp"
            type="image/webp"
          />
           
          <source
            alt="Levekårs&shy;indikatorer"
            srcset="/img/levekaar.png"
            type="image/png"
          />
           
          <img
            alt="Levekårs&shy;indikatorer"
            aria-hidden="true"
            class="oslo__img"
            src="/img/levekaar.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(244, 206, 138);"
        >
          Levekår
        </span>
         
        <span
          class="oslo__topic"
        >
          Levekårs­indikatorer
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/eierform"
        id="eierform"
      >
        <picture>
          <source
            alt="Eierform"
            srcset="/img/eieform.webp"
            type="image/webp"
          />
           
          <source
            alt="Eierform"
            srcset="/img/eieform.png"
            type="image/png"
          />
           
          <img
            alt="Eierform"
            aria-hidden="true"
            class="oslo__img"
            src="/img/eieform.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(199, 247, 201);"
        >
          Boforhold
        </span>
         
        <span
          class="oslo__topic"
        >
          Eierform
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/bygningstyper"
        id="bygningstyper"
      >
        <picture>
          <source
            alt="Bygningstyper"
            srcset="/img/bygningstyper.webp"
            type="image/webp"
          />
           
          <source
            alt="Bygningstyper"
            srcset="/img/bygningstyper.png"
            type="image/png"
          />
           
          <img
            alt="Bygningstyper"
            aria-hidden="true"
            class="oslo__img"
            src="/img/bygningstyper.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(199, 247, 201);"
        >
          Boforhold
        </span>
         
        <span
          class="oslo__topic"
        >
          Bygningstyper
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/boligpriser"
        id="boligpriser"
      >
        <picture>
          <source
            alt="Boligpriser"
            srcset="/img/boligpriser.webp"
            type="image/webp"
          />
           
          <source
            alt="Boligpriser"
            srcset="/img/boligpriser.png"
            type="image/png"
          />
           
          <img
            alt="Boligpriser"
            aria-hidden="true"
            class="oslo__img"
            src="/img/boligpriser.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(199, 247, 201);"
        >
          Boforhold
        </span>
         
        <span
          class="oslo__topic"
        >
          Boligpriser
        </span>
      </a>
      <a
        class="main-container__item"
        href="#/bydel/gamleoslo/rom-per-person"
        id="rom-per-person"
      >
        <picture>
          <source
            alt="Rom per person"
            srcset="/img/trangboddhet.webp"
            type="image/webp"
          />
           
          <source
            alt="Rom per person"
            srcset="/img/trangboddhet.png"
            type="image/png"
          />
           
          <img
            alt="Rom per person"
            aria-hidden="true"
            class="oslo__img"
            src="/img/trangboddhet.png"
          />
        </picture>
         
        <span
          class="oslo__category"
          style="background-color: rgb(199, 247, 201);"
        >
          Boforhold
        </span>
         
        <span
          class="oslo__topic"
        >
          Rom per person
        </span>
      </a>
    </div>
  </div>
</div>
`;
